File: 198-house-robber/house-robber.cpp
class Solution {
public:
    int helper(vector<int>& nums, int i, vector<int>& dp){
        if(i >= nums.size()) return 0;

        if(dp[i] != -1) return dp[i];

        int pick = nums[i] + helper(nums, i+2,dp);
        int unpick = helper(nums, i+1, dp);

        return dp[i] = max(pick, unpick);
    }
    int rob(vector<int>& nums) {
        vector<int> dp(nums.size(), -1);
        return helper(nums, 0, dp);
    }
};

File: 198-house-robber/house-robber.py
class Solution:
    def helper(self, nums: List[int], i: int, dp:List[int]) -> int:
        if i >= len(nums):
            return 0

        if dp[i] != -1:
            return dp[i]

        pick = nums[i] + self.helper(nums, i+2, dp)
        unpick = self.helper(nums, i+1, dp)

        dp[i] = max(pick, unpick)

        return dp[i]
    def rob(self, nums: List[int]) -> int:
        dp = [-1] * (len(nums)+1)
        return self.helper(nums, 0, dp)